core/trino-main/src/test/java/io/trino/sql/query/TestSessionProperties.java

```diff
@@ -0,0 +1,155 @@
+                .isEqualTo(Map.of("optimize_hash_generation", "true"));
+
+        assertThat(analyze("WITH SESSION optimize_hash_generation = CAST('true' AS boolean) SELECT 1").getSystemProperties())
+                .isEqualTo(Map.of("optimize_hash_generation", "true"));
+
+        assertThatThrownBy(() -> analyze("WITH SESSION optimize_hash_generation = true, optimize_hash_generation = false SELECT 1"))
+                .hasMessageContaining("line 1:47: Session property optimize_hash_generation already set");
+    }
+
+    @Test
+    void testCatalogSessionProperty()
+    {
+        assertThatThrownBy(() -> analyze("WITH SESSION test.invalid_key = 'invalid_value' SELECT 1"))
+                .hasMessageContaining("line 1:14: Session property test.invalid_key does not exist");
+
+        assertThatThrownBy(() -> analyze("WITH SESSION test.catalog_property = true SELECT 1"))
+                .hasMessageContaining("Unable to set session property 'test.catalog_property' to 'true': Cannot cast type boolean to varchar");
+
+        assertThat(analyze("WITH SESSION test.catalog_property = 'true' SELECT 1").getCatalogProperties("test"))
+                .isEqualTo(Map.of("catalog_property", "true"));
+
+        assertThat(analyze("WITH SESSION test.catalog_property = CAST(true AS varchar) SELECT 1").getCatalogProperties("test"))
+                .isEqualTo(Map.of("catalog_property", "true"));
+
+        assertThatThrownBy(() -> analyze("WITH SESSION test.catalog_property = 'true', test.catalog_property = 'false' SELECT 1").getCatalogProperties("test"))
+                .hasMessageContaining("line 1:46: Session property test.catalog_property already set");
+    }
+
+    @Test
+    void testInvalidSessionProperty()
+    {
+        assertThatThrownBy(() -> analyze("WITH SESSION test.schema.invalid_key = 'invalid_value' SELECT 1"))
+                .hasMessageContaining("line 1:14: Invalid session property 'test.schema.invalid_key'");
+    }
+
+    private static Session analyze(@Language("SQL") String statement)
+    {
+        TransactionManager transactionManager = new TestingTransactionManager();
+        PlannerContext plannerContext = plannerContextBuilder()
+                .withMetadata(new MockMetadata())
+                .withTransactionManager(transactionManager)
+                .build();
+
+        return transaction(transactionManager, plannerContext.getMetadata(), new AllowAllAccessControl())
+                .execute(testSession(), transactionSession -> {
+                    SessionPropertyInterpreter evaluator = new SessionPropertyInterpreter(plannerContext, new AllowAllAccessControl(), SESSION_PROPERTY_MANAGER);
+                    QueryPreparer.PreparedQuery preparedQuery = new QueryPreparer.PreparedQuery(SQL_PARSER.createStatement(statement), ImmutableList.of(), Optional.empty());
+                    return evaluator.getSessionPropertiesApplier(preparedQuery).apply(transactionSession);
+                });
+    }
```

Comment:
Why do we need a custom analyzer? It would be ideal to use the existing runner and assertion frameworks as much as possible (maybe with some small extensions). This kind of setup is hard to maintain and evolve, as it's sensitive to changes to internal APIs.